datajunction-server/datajunction_server/api/helpers.py

```diff
@@ -616,3 +622,114 @@ def validate_orderby(
                 "specified in the metrics or dimensions"
             ),
         )
+
+
+def find_existing_cube(
+    session: Session,
+    metric_columns: List[Column],
+    dimension_columns: List[Column],
+) -> Optional[NodeRevision]:
+    """
+    Find an existing cube with these metrics and dimensions, if any
+    """
+    element_names = [col.name for col in (metric_columns + dimension_columns)]
+    statement = select(NodeRevision)
+    for name in element_names:
+        statement = statement.filter(
+            NodeRevision.cube_elements.any(Column.name == name),  # type: ignore  # pylint: disable=no-member
+        )
+
+    existing_cubes = session.exec(statement).unique().all()
+    for cube in existing_cubes:
+        return cube
+    return None
+
+
+def build_sql_for_multiple_metrics(  # pylint: disable=too-many-arguments,too-many-locals
+    session: Session,
+    metrics: List[str],
+    dimensions: List[str],
+    filters: List[str] = None,
+    orderby: List[str] = None,
+    limit: Optional[int] = None,
+    engine_name: Optional[str] = None,
+    engine_version: Optional[str] = None,
+) -> Tuple[TranslatedSQL, Engine, Catalog]:
+    """
+    Build SQL for multiple metrics. Used by both /sql and /data endpoints
+    """
+    if not filters:
+        filters = []
+    if not orderby:
+        orderby = []
+
+    metric_columns, metric_nodes, _, dimension_columns, _ = validate_cube(
+        session,
+        metrics,
+        dimensions,
+    )
+    leading_metric_node = get_node_by_name(session, metrics[0])
+    available_engines = leading_metric_node.current.catalog.engines
+
+    # Try to find a built cube that already has the given metrics and dimensions
+    # The cube needs to have a materialization configured and an availability state
+    # posted in order for us to use the materialized datasource directly
+    cube = find_existing_cube(session, metric_columns, dimension_columns)
+    if cube and cube.materializations and cube.availability:
```

Comment:
nit: I wonder if it makes sense to add this condition to the `find_existing_cube()` method itself... perhaps not, but then I feel like we could add a wrapper on that one called `find_materialized_cube()` and have this condition in there for reuse.

Reply:
Agreed, I think it would be a good idea to add an option to look for materialized cubes. I added a parameter to `find_existing_cube` called `materialized` so that we can filter for materialized cubes if desired.